src/raid.cpp

```diff
@@ -0,0 +1,911 @@
+{
+    assert(tempUrls.size() == RAIDPARTS || tempUrls.size() == 1);
+    assert(0 <= resumepos && resumepos <= readtopos && readtopos <= filesize);
+    assert(!raidKnown);
+
+    tempurls = tempUrls;
+
+    is_raid = tempurls.size() == RAIDPARTS;
+    raidKnown = true;
+    fullfilesize = filesize;
+    deliverlimitpos = readtopos;
+    acquirelimitpos = deliverlimitpos + RAIDLINE - 1;
+    acquirelimitpos -= acquirelimitpos % RAIDLINE;
+    acquirelimitpos = std::min<m_off_t>(acquirelimitpos, fullfilesize);
+    outputfilepos = resumepos;
+    if (is_raid)
+    {
+        raidpartspos = resumepos / (RAIDPARTS - 1);
+        raidpartspos -= raidpartspos % RAIDSECTOR;
+        resumewastedbytes = size_t(outputfilepos - raidpartspos * (RAIDPARTS - 1));
+        outputfilepos -= resumewastedbytes;  // we'll skip over these bytes on the first output
+        for (int i = RAIDPARTS; i--; )
+        {
+            raidrequestpartpos[i] = raidpartspos;
+        }
+
+        // How much buffer space can we use.  Assuming two chunk sets incoming, one outgoing
+        raidLinesPerChunk = unsigned(maxRequestSize / (RAIDPARTS * 3 * RAIDSECTOR));
+        raidLinesPerChunk -= raidLinesPerChunk % 1024;
+        raidLinesPerChunk = std::min<unsigned>(raidLinesPerChunk, 64 * 1024);
+        raidLinesPerChunk = std::max<unsigned>(raidLinesPerChunk, 8 * 1024);
+
+        unusedRaidConnection = g_faultyServers.selectWorstServer(tempurls);
+    }
+
+    DEBUG_TEST_HOOK_RAIDBUFFERMANAGER_SETISRAID(this)
+}
+
+void RaidBufferManager::updateUrlsAndResetPos(const std::vector<std::string>& tempUrls)
+{
+    // A request to restart from whereever we got to, with new URLs.
+    // the old requested-to pos is not valid anymore, as one or more Http requests failed or were abandoned
+    assert(tempurls.size() == tempUrls.size());
+    if (tempurls.size() == tempUrls.size())
+    {
+        tempurls = tempUrls;
+        if (isRaid())
+        {
+            for (unsigned i = RAIDPARTS; i--; )
+            {
+                std::deque<FilePiece*>& connectionpieces = raidinputparts[i];
+                transferPos(i) = connectionpieces.empty() ? raidpartspos : connectionpieces.back()->pos + connectionpieces.back()->buf.datalen();
+            }
+        }
+        else
+        {
+            transferPos(0) = outputfilepos;  // if there is any data waiting in asyncoutputbuffers this value is alreday ahead of it
+        }
+    }
+}
+
+bool RaidBufferManager::isRaid()
+{
+    assert(raidKnown);
+    return is_raid;
+}
+
+std::string RaidBufferManager::emptyReturnString;
+
+const std::string& RaidBufferManager::tempURL(unsigned connectionNum)
+{
+    if (isRaid())
+    {
+        assert(connectionNum < tempurls.size());
+        return tempurls[connectionNum];
+    }
+    else if (!tempurls.empty())
+    {
+        return tempurls[0];
+    }
+    else
+    {
+        return emptyReturnString;
```

Comment:
Is it really safe to return an empty string here? Could HttpReqXfer::prepare() be called with an empty string? Not sure about it, just wondering.

Reply:
the empty string there is just so that the function can return something legal (by reference) if the vector is empty.  That should never occur since the function shouldn't be used unless we got a reply for URLs to upload/download to/from.  I've added an assert in to clarify (and verify) that.  The return is by reference for efficiency, no need to copy strings in the normal case.  I've also changed it to a normal class member instead of a static, just in case of that initialization sequence issue.   thanks!